File: IPaURLResourceUI/IPaURLResourceUI/IPaURLResourceUI.m
//
//  IPaURLResourceUI.m
//  IPaURLResourceUI
//
//  Created by IPa Chen on 2014/10/10.
//  Copyright (c) 2014年 IPaPa. All rights reserved.
//

#import "IPaURLResourceUI.h"
@interface IPaURLResourceUI() <NSURLSessionDelegate>
@end
@implementation IPaURLResourceUI
- (instancetype)init
{
    self  = [super init];
    self.filterNSNull = NO;
    return self;
}


-(NSURLSessionConfiguration*)sessionConfiguration;
{
    if (_sessionConfiguration == nil) {
        _sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
    }
    return _sessionConfiguration;
}
-(NSURLSession*)urlSession
{
    if (_urlSession == nil) {
        _urlSession = [NSURLSession sessionWithConfiguration:self.sessionConfiguration delegate:self delegateQueue:[NSOperationQueue mainQueue]];
    }
    return _urlSession;
}
-(NSURLSessionDataTask*)apiGet:(NSString*)api param:(NSDictionary *)param onComplete:(IPaURLResourceUISuccessHandler)complete failure:(void (^)(NSError*))failure
{

    NSString *apiURL = [self.baseURL stringByAppendingString:api];
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];

    [request setHTTPMethod:@"GET"];

    if ([param count] > 0) {
        apiURL = [apiURL stringByAppendingString:@"?"];
        NSUInteger count = 0;
        for (NSString* key in param.allKeys) {
            NSString* value = param[key];
            
            apiURL = [apiURL stringByAppendingFormat:(count > 0)?@"&%@=%@":@"%@=%@",key,value];
            count++;
            
        }
        
    }
    apiURL = [apiURL stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    NSURL *url = [NSURL URLWithString:apiURL];
    [request setURL:url];
    return [self apiWithRequest:request onComplete:complete failure:failure];
}

-(NSURLSessionDataTask*)apiPost:(NSString*)api param:(NSDictionary *)param onComplete:(IPaURLResourceUISuccessHandler)complete failure:(void (^)(NSError*))failure
{
    return [self api:api method:@"POST" paramInBody:param onComplete:complete failure:failure];
}
-(NSURLSessionDataTask*)apiPut:(NSString*)api param:(NSDictionary *)param onComplete:(IPaURLResourceUISuccessHandler)complete failure:(void (^)(NSError*))failure
{
    return [self api:api method:@"PUT" paramInBody:param onComplete:complete failure:failure];
}
-(NSURLSessionDataTask*)api:(NSString*)api method:(NSString*)method paramInBody:(NSDictionary *)paramInBody onComplete:(IPaURLResourceUISuccessHandler)complete failure:(void (^)(NSError*))failure
{
    NSString *apiURL = [self.baseURL stringByAppendingString:api];
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    [request setURL:[NSURL URLWithString:apiURL]];
    [request setHTTPMethod:method];
    
    [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"content-type"];
    NSUInteger count = 0;
    NSString *postString;
    for (NSString* key in paramInBody.allKeys) {
        NSString* value = paramInBody[key];
        
        postString = [postString stringByAppendingFormat:(count > 0)?@"&%@=%@":@"%@=%@",key,value];
        count++;
        
    }
    [request setHTTPBody:[postString dataUsingEncoding:NSUTF8StringEncoding]];
    
    //    postString = [postString stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    return [self apiWithRequest:request onComplete:complete failure:failure];
}
-(NSURLSessionUploadTask*)api:(NSString*)api uploadWithMethod:(NSString*)method contentType:(NSString*)contentType data:(NSData*)data onComplete:(IPaURLResourceUISuccessHandler)complete failure:(void (^)(NSError*))failure
{
    NSString *apiURL = [self.baseURL stringByAppendingString:api];
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    [request setURL:[NSURL URLWithString:apiURL]];
    [request setHTTPMethod:method];
    
    [request setValue:contentType forHTTPHeaderField:@"content-type"];
//    NSString *dataLength = [NSString stringWithFormat:@"%ld", (unsigned long)[data length]];
//    [request setValue:dataLength forHTTPHeaderField:@"Content-Length"];
//    [request setHTTPBody:data];
    
    NSURLSessionUploadTask *task = [self.urlSession uploadTaskWithRequest:request fromData:data completionHandler:^(NSData* responseData,NSURLResponse* response,NSError* error){
        if (error != nil) {
            if (failure) {
                failure(error);
            }
            return;
        }
        NSError *jsonError;
#ifdef DEBUG

        NSString *retString = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
        NSLog(@"IPaURLResourceUI return string :%@",retString);
#endif
        id jsonData = [NSJSONSerialization JSONObjectWithData:responseData options:0 error:&jsonError];
        if (jsonError != nil) {
            if (failure) {
                failure(error);
            }
            return;
        }
        
        if (complete) {
            if (self.filterNSNull) {
                jsonData = [self removeNSNullDataFromObject:jsonData];
            }
            complete(response,jsonData);
        }
        
    }];
    
    [task resume];
    return task;
    

}
-(NSURLSessionUploadTask*)api:(NSString*)api uploadMultipartFormDataWithMethod:(NSString*)method bodyData:(IPaURLMultipartFormData*)bodyData onComplete:(IPaURLResourceUISuccessHandler)complete failure:(void (^)(NSError*))failure
{

    NSString *contentType = [NSString stringWithFormat:@"multipart/form-data; boundary=%@", bodyData.boundary];
    NSData* data = [bodyData createBodyData];

    
    return [self api:api uploadWithMethod:method contentType:contentType data:data onComplete:complete failure:failure];
}
-(NSURLSessionUploadTask*)apiPut:(NSString*)api contentType:(NSString*)contentType postData:(NSData*)postData onComplete:(IPaURLResourceUISuccessHandler)complete failure:(void (^)(NSError*))failure
{
    return [self api:api uploadWithMethod:@"PUT" contentType:contentType data:postData onComplete:complete failure:failure];
}
-(NSURLSessionUploadTask*)apiPost:(NSString*)api contentType:(NSString*)contentType postData:(NSData*)postData onComplete:(IPaURLResourceUISuccessHandler)complete failure:(void (^)(NSError*))failure
{
    return [self api:api uploadWithMethod:@"POST" contentType:contentType data:postData onComplete:complete failure:failure];
    
}
-(NSURLSessionDataTask*) apiWithRequest:(NSURLRequest*)request  onComplete:(IPaURLResourceUISuccessHandler)complete failure:(void (^)(NSError*))failure
{
    NSURLSessionDataTask *task = [self.urlSession dataTaskWithRequest:request completionHandler:^(NSData* responseData,NSURLResponse* response,NSError* error){
        if (error != nil) {
            if (failure) {
                failure(error);
            }
            return;
        }
        NSError *jsonError;
        id jsonData = [NSJSONSerialization JSONObjectWithData:responseData options:0 error:&jsonError];
        if (jsonError != nil) {
            if (failure) {
                failure(error);
            }
            return;
        }
        
        if (complete) {
            
            if (self.filterNSNull) {
                jsonData = [self removeNSNullDataFromObject:jsonData];
            }
            complete(response,jsonData);
        }
        
    }];
    
    [task resume];
    return task;
}
- (void)URLSession:(NSURLSession *)session didBecomeInvalidWithError:(NSError *)error
{
    NSLog(@"%@",error);
}
- (void)URLSessionDidFinishEventsForBackgroundURLSession:(NSURLSession *)session {
}
- (id)removeNSNullDataFromObject:(id)object
{
    if ([object isKindOfClass:[NSDictionary class]]) {
        return [self removeNSNullDataFromDictionary:object];
    }
    else if ([object isKindOfClass:[NSArray class]]) {
        return [self removeNSNullDataFromArray:object];
    }
    return object;
}
- (NSDictionary *)removeNSNullDataFromDictionary:(NSDictionary*)dictionary
{
    NSMutableDictionary *mDict = [@{} mutableCopy];
    for (NSString *key in dictionary.allKeys) {
        id value = dictionary[key];
        if ([value isEqual:[NSNull null]]) {
            continue;
        }
        else if([value isKindOfClass:[NSDictionary class]]) {
            value = [self removeNSNullDataFromDictionary:value];
        }
        else if ([value isKindOfClass:[NSArray class]]) {
            value = [self removeNSNullDataFromArray:value];
        }
        mDict[key] = value;
    }
    return mDict;
}
- (NSArray *)removeNSNullDataFromArray:(NSArray*)array
{
    NSMutableArray *mArray = [@[] mutableCopy];
    for (id value in array) {
        id newValue = value;
        if ([value isEqual:[NSNull null]]) {
            continue;
        }
        else if([value isKindOfClass:[NSDictionary class]]) {
            newValue = [self removeNSNullDataFromDictionary:value];
        }
        else if ([value isKindOfClass:[NSArray class]]) {
            newValue = [self removeNSNullDataFromArray:value];
        }
        [mArray addObject:newValue];
    }
    return mArray;
}
@end

File: IPaURLResourceUI/IPaURLResourceUI/IPaURLResourceUI.swift
//
//  IPaURLResourceUI.swift
//  IPaURLResourceUI
//
//  Created by IPa Chen on 2015/6/12.
//  Copyright (c) 2015年 IPaPa. All rights reserved.
//

import Foundation
public typealias IPaURLResourceUISuccessHandler = ((URLResponse?,Any?) -> ())!
public typealias IPaURLResourceUIFailHandler = ((Error) -> ())!
open class IPaURLResourceUI : NSObject,URLSessionDelegate {
    var baseURL:String! = ""
    open var removeNSNull:Bool = true
    lazy var sessionConfiguration:URLSessionConfiguration = URLSessionConfiguration.default
    lazy var urlSession:Foundation.URLSession = Foundation.URLSession(configuration: self.sessionConfiguration, delegate: self, delegateQueue: OperationQueue.main)
    func urlStringForAPI(_ api:String) -> String {
        return self.baseURL + api
    }
    func urlStringForGETAPI(_ api:String!, param:[String:Any]?) -> String! {
        var apiURL = self.baseURL + api
        
        if let params = param {
            apiURL = apiURL + "?"
            var count = 0
            
            for key in params.keys {
                apiURL = apiURL + ((count > 0) ? "&":"") + "\(key)=\(params[key]!)"
                count += 1
            }
        }
        
        return ((apiURL as NSString).addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed))!
        
        
        
        
    }
    func apiWithRequest(_ request:URLRequest,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionDataTask {
        
        let task = urlSession.dataTask(with: request, completionHandler: { (responseData,response,error) in
            if let error = error {
                failure(error)
                return
            }
            var jsonData:Any?
            do {
                if let responseData = responseData {
                    jsonData = try JSONSerialization.jsonObject(with: responseData, options: JSONSerialization.ReadingOptions())
                }
            } catch let error as NSError {
                jsonData = nil
                #if DEBUG
                if let responseString = String(data: responseData!, encoding: String.Encoding.utf8) {
                    print(responseString)
                }
                #endif
                
                failure(error)
                
                return
            } catch {
                fatalError()
            }
            if self.removeNSNull {
                jsonData = self.removeNSNullDataFromObject(jsonData!)
            }
            complete(response,jsonData)
        })
        
        task.resume()
        
        return task
    }
    
    open func apiGet(_ api:String ,param:[String:Any]?,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionDataTask
        
    {
        let apiURL = urlStringForGETAPI(api, param: param)
        let request = NSMutableURLRequest()
        request.httpMethod = "GET"
        request.url = URL(string: apiURL!)
        
        return apiWithRequest(request as URLRequest,complete:complete,failure:failure)
    }
    func apiPost(_ api:String , param:[String:Any]?,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionDataTask
    {
        return apiPerform(api,method:"POST",paramInBody:param,complete:complete,failure:failure)
    }
    
    
    func apiPut(_ api:String, param:[String:Any]?,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionDataTask {
        
        return apiPerform(api ,method:"PUT", paramInBody:param,complete:complete,failure:failure)
    }
    func apiUpload(_ api:String,method:String,json:Any,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: json, options: JSONSerialization.WritingOptions(rawValue: 0))
            _ = apiUpload(api, method: method, headerParam: ["content-type":"application/json"], data: jsonData, complete: complete, failure: failure)
            
        }
        catch let error as NSError {
            failure(error)
            return
        }
        
        
        
    }
    func apiPerform(_ api:String,method:String,paramInHeader:[String:String]?,paramInBody:[String:Any]?,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionDataTask
    {
        let method = method.uppercased()
        let apiURL = (method == "GET") ? urlStringForGETAPI(api, param: paramInBody) :  urlStringForAPI(api)
        let request = NSMutableURLRequest()
        if let param = paramInBody , method != "GET" {
            var count = 0
            var postString = ""
            
            for key in param.keys {
                var value = "\(param[key]!)"
                let characterSet = NSMutableCharacterSet.alphanumeric()
                characterSet.addCharacters(in: "-._~")
                value = (value.addingPercentEncoding(withAllowedCharacters: characterSet as CharacterSet))!
                
                postString = postString + ((count > 0) ? "&" : "") + "\(key)=\(value)"
                count += 1
                
            }
            request.httpBody = postString.data(using: String.Encoding.utf8, allowLossyConversion: false)
            
        }
        
        request.url = URL(string: apiURL!)
        request.httpMethod = method
        if let param = paramInHeader {
            for key in param.keys {
                request.setValue(param[key]!,forHTTPHeaderField: key)
            }
        }
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "content-type")
        
        return apiWithRequest(request as URLRequest,complete:complete,failure:failure)
        
        
    }
    func apiPerform(_ api:String,method:String,paramInBody:[String:Any]?,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionDataTask {
        return apiPerform(api, method: method, paramInHeader: nil, paramInBody: paramInBody, complete: complete, failure: failure)
        
    }
    func apiUpload(_ api:String,method:String,headerParam:[String:String],data:Data,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionUploadTask {
        let apiURL = urlStringForAPI(api)
        let request = NSMutableURLRequest()
        request.url = URL(string: apiURL)
        request.httpMethod = method
        
        for key in headerParam.keys {
            request.setValue(headerParam[key]!, forHTTPHeaderField: key)
            
        }
        //        [request setValue:contentType forHTTPHeaderField:@"content-type"]
        //    NSString *dataLength = [NSString stringWithFormat:@"%ld", (unsigned long)[data length]]
        //    [request setValue:dataLength forHTTPHeaderField:@"Content-Length"]
        //    [request setHTTPBody:data]
        let task = urlSession.uploadTask(with: request as URLRequest, from: data, completionHandler: {
            (responseData,response,error) -> Void in
            if let error = error {
                failure(error)
                return
            }
            
            #if DEBUG
                if let responseData = responseData {
                    let retString = NSString(data: responseData, encoding: String.Encoding.utf8.rawValue)
                    print("IPaURLResourceUI return string :\(retString)")
                }
            #endif
            var jsonData:Any?
            do {
                if let responseData = responseData {
                    jsonData = try JSONSerialization.jsonObject(with: responseData, options: JSONSerialization.ReadingOptions())
                }
            } catch let error as NSError {
                failure(error)
                return
            } catch {
                fatalError()
            }
            if self.removeNSNull {
                jsonData = self.removeNSNullDataFromObject(jsonData!)
            }
            complete(response,jsonData)
            
            
        })
        
        task.resume()
        return task
    }
    func apiUpload(_ api:String,method:String,multiPartFormData:IPaURLMultipartFormData,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionUploadTask {
        let contentType = "multipart/form-data boundary=\(multiPartFormData.boundary)"
        multiPartFormData.endOfBodyData()
        let data = multiPartFormData.data
        return apiUpload(api, method: method, headerParam: ["content-type":contentType], data: data as Data, complete: complete, failure: failure)
    }
    func apiPut(_ api:String,contentType:String,postData:Data,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionUploadTask {
        return apiUpload(api, method: "PUT", headerParam: ["content-type":contentType], data: postData, complete: complete, failure: failure)
    }
    func apiPost(_ api:String,contentType:String,postData:Data,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionUploadTask {
        return apiUpload(api, method: "POST", headerParam: ["content-type":contentType], data: postData, complete: complete, failure: failure)
    }
    func apiDelete(_ api:String,contentType:String,postData:Data,complete:IPaURLResourceUISuccessHandler,failure:IPaURLResourceUIFailHandler) -> URLSessionUploadTask {
        return apiUpload(api, method: "DELETE", headerParam: ["content-type":contentType], data: postData, complete: complete, failure: failure)
    }
    
    
    
    // MARK:NSURLSessionDelegate
    open func urlSession(_ session: URLSession, didBecomeInvalidWithError error: Error?) {
        IPaLog("\(error)")
    }
    open func urlSessionDidFinishEvents(forBackgroundURLSession session: URLSession) {
        
    }
    
    // MARK: helper method
    func removeNSNullDataFromObject(_ object:Any) -> Any
    {
        if let dictObject = object as? [String:Any] {
            return removeNSNullDataFromDictionary(dictObject) as Any
        }
        else if let arrayValue = object as? [Any] {
            return removeNSNullDataFromArray(arrayValue) as Any
        }
        return object;
    }
    func removeNSNullDataFromDictionary(_ dictionary:[String:Any?]) -> [String:Any?]
    {
        var mDict = [String:Any]()
        
        for key in dictionary.keys {
            var value = dictionary[key]
            if let _ = value as? NSNull {
                continue;
            }
            else if let dictValue = value as? [String:Any] {
                value = removeNSNullDataFromDictionary(dictValue) as Any?
            }
            else if let arrayValue = value as? [Any] {
                value = removeNSNullDataFromArray(arrayValue) as Any?
            }
            mDict[key] = value!;
        }
        return mDict;
    }
    func removeNSNullDataFromArray(_ array:[Any]) -> [Any]
    {
        var mArray = [Any]()
        for value in array {
            var newValue:Any = value;
            if value is NSNull {
                continue;
            }
            else if let dictValue = value as? [String:Any] {
                newValue = removeNSNullDataFromDictionary(dictValue) as Any
            }
            else if let arrayValue = value as? [Any] {
                newValue = removeNSNullDataFromArray(arrayValue) as Any
            }
            mArray.append(newValue)
            
        }
        return mArray;
    }
    
}


